### PR TITLE
RAC-38: feat: add permission on quantified associations

### PR DESCRIPTION
**Pull request by StevenVAIDIE**
**Description (for Contributor and Core Developer)**
See https://github.com/akeneo/pim-enterprise-dev/pull/9064

This PR add methods in the new `QuantifiedAssociations` object to filter products & product models based on identifiers.
This is needed & used by the EE permissions system, as the EE permissions filter unauthorized products from associations.
As a consequence, the connector read models needed to be updated too.

Since we check permissions against an user, a few acceptance tests without authenticated user were now failing.
I would rather have `Given an authentified user` in a few acceptance tests than add exceptions everywhere where the user could be null, which should never be the case in a real life scenario.

**Definition Of Done (for Core Developer only)**

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | Todo
| Added legacy Behats               | Todo
| Added acceptance tests            | Todo
| Added integration tests           | Todo
| Changelog updated                 | Todo
| Review and 2 GTM                  | Todo
| Micro Demo to the PO (Story only) | Todo
| Migration script                  | -
| Tech Doc                          | -

`Todo`: Pending / Work in progress
`OK`: Done / Validated
`-`: Not needed

### CODE REVIEW

File: src/Akeneo/Pim/Enrichment/Component/Product/Connector/ReadModel/ConnectorProduct.php [hunk 1]
```diff
@@ -225,6 +225,32 @@ public function associatedProductModelCodes(): array
         return !empty($associatedProductModels) ? array_unique(array_merge(...$associatedProductModels)) : [];
     }
 
+    public function associatedWithQuantityProductIdentifiers()
```
Comment (juliensnz) [inline src/Akeneo/Pim/Enrichment/Component/Product/Connector/ReadModel/ConnectorProduct.php:228]: I don't find this name very clear

File: src/Akeneo/Pim/Enrichment/Component/Product/Connector/ReadModel/ConnectorProduct.php [hunk 2]
```diff
@@ -254,6 +280,70 @@ public function filterAssociatedProductModelsByProductModelCodes(array $productM
...
+                    return in_array($quantifiedLink['identifier'], $productIdentifiersToFilter);
+                }
+            );
+
+            $filteredQuantifiedAssociations[$associationType]['products'] = array_values($filteredProductQuantifiedAssociations);
+            $filteredQuantifiedAssociations[$associationType]['product_models'] = $quantifiedAssociation['product_models'];
+        }
+
+        return new self(
+            $this->id,
```
Comment (juliensnz) [inline src/Akeneo/Pim/Enrichment/Component/Product/Connector/ReadModel/ConnectorProduct.php:331]: we re-create a product from scratch?

File: src/Akeneo/Pim/Enrichment/Component/Product/Connector/ReadModel/ConnectorProductModel.php [hunk 3]
```diff
@@ -164,6 +164,32 @@ public function associatedProductModelCodes(): array
         return !empty($associatedProductModels) ? array_unique(array_merge(...$associatedProductModels)) : [];
     }
 
+    public function associatedWithQuantityProductIdentifiers()
```
Comment (juliensnz) [inline src/Akeneo/Pim/Enrichment/Component/Product/Connector/ReadModel/ConnectorProductModel.php:167]: duplicated code?

**Comment by tseho** [comment]
Sorry about that, I will update the description.
This PR is only the CE part of the permissions PR.

File: src/Akeneo/Pim/Enrichment/Component/Product/Connector/ReadModel/ConnectorProduct.php (hunk 2 above)
Comment (tseho) [reply src/Akeneo/Pim/Enrichment/Component/Product/Connector/ReadModel/ConnectorProduct.php:331]: It's a ConnectorProduct, it's only a immutable read model and it's done exactly like this in the other methods, see on master: https://github.com/akeneo/pim-community-dev/blob/e869b0ce1358277f3e448473196d914eab500951/src/Akeneo/Pim/Enrichment/Component/Product/Connector/ReadModel/ConnectorProduct.php#L181

File: src/Akeneo/Pim/Enrichment/Component/Product/Connector/ReadModel/ConnectorProduct.php (hunk 1 above)
Comment (tseho) [reply src/Akeneo/Pim/Enrichment/Component/Product/Connector/ReadModel/ConnectorProduct.php:228]: I kinda agree with you, do you have a suggestion ?

File: src/Akeneo/Pim/Enrichment/Component/Product/Connector/ReadModel/ConnectorProductModel.php (hunk 3 above)
Comment (tseho) [reply src/Akeneo/Pim/Enrichment/Component/Product/Connector/ReadModel/ConnectorProductModel.php:167]: It's for ProductModel this time, and these 2 models are built this.
It was already designed like this, IMHO Steven did the right thing by staying consistent with the existing way of doing these things.

Comment (juliensnz) [reply src/Akeneo/Pim/Enrichment/Component/Product/Connector/ReadModel/ConnectorProductModel.php:167]: yes you are right

File: src/Akeneo/Pim/Enrichment/Component/Product/Connector/ReadModel/ConnectorProduct.php (hunk 2 above)
Comment (juliensnz) [reply src/Akeneo/Pim/Enrichment/Component/Product/Connector/ReadModel/ConnectorProduct.php:331]: ok, no pb :)

File: src/Akeneo/Pim/Enrichment/Component/Product/Connector/ReadModel/ConnectorProduct.php (hunk 1 above)
Comment (juliensnz) [reply src/Akeneo/Pim/Enrichment/Component/Product/Connector/ReadModel/ConnectorProduct.php:228]: getQuantifiedAssociationProductIdentifiers?

**Pull request merged**